Frameworks/Foundation/NSData.mm

```diff
@@ -396,42 +388,29 @@ - (instancetype)initWithContentsOfFile:(NSString*)filename options:(NSDataReadin
         return nil;
     }
 
-    char* fname = (char*)[filename UTF8String];
-
-    TraceVerbose(TAG, L"NSData extended-opening %hs", fname);
-    EbrFile* fpIn = EbrFopen(fname, "rb");
-    if (fpIn) {
-        auto closeFile = wil::ScopeExit([&]() { EbrFclose(fpIn); });
-
-        EbrFseek(fpIn, 0, SEEK_END);
-        size_t length = EbrFtell(fpIn);
-        EbrFseek(fpIn, 0, SEEK_SET);
-
-        if (length) {
-            // The created NSData takes ownership of freeing this
-            woc::unique_iw<uint8_t> copiedBytes(static_cast<uint8_t*>(IwMalloc(length)));
-            if (!copiedBytes.get()) {
-                if (error) {
-                    *error = [NSError errorWithDomain:@"NSData" code:100 userInfo:nil];
-                }
-                [self release];
-                return nil;
-            }
-
-            size_t fileLength = EbrFread(copiedBytes.get(), 1, length, fpIn);
-            self = [self initWithBytesNoCopy:copiedBytes.release() length:fileLength];
-        }
+    NSInputStream* inputStream = [NSInputStream inputStreamWithFileAtPath:filename];
+    [inputStream open];
 
-    } else {
-        TraceVerbose(TAG, L"NSData couldn't open %hs for read (extended)", fname);
+    if (NSStreamStatusOpen != inputStream.streamStatus) {
+        TraceVerbose(TAG, L"NSData couldn't open %hs for read (extended)", filename);
         if (error) {
             *error = [NSError errorWithDomain:@"NSData" code:100 userInfo:nil];
         }
         [self release];
         return nil;
     }
 
-    return self;
+    auto closeFile = wil::ScopeExit([&]() { [inputStream close]; });
+
+    uint8_t byteBuffer[4096] = {};
+    NSMutableData* tempData = [NSMutableData data];
+
+    while ([inputStream hasBytesAvailable]) {
+        auto result = [inputStream read:byteBuffer maxLength:_countof(byteBuffer)];
```

Comment:
increaseLength isn't really desirable here. Ideally I'd be able to read the chunks directly in and return the same data. If the length is wrong I can't return that data and need to manually reconstruct it with the right length that I keep track of. appendBytes is morally equivalent of what you are suggesting but it adjusts the capacity instead of length at the expense of the temp buffer (which is only allocated once and is stack space so its "cheap" in the grand scheme of things). Basically this is letting the NSData append impl be smart about growth etc which seems better than me trying to be smart (since I'm not very). I'll look for other ways to avoid extra copies though.

---
In reply to: [83929862](https://github.com/Microsoft/WinObjC/pull/1175#discussion_r83929862) [](ancestors = 83929862)